### ISSUE TITLE
Endless loop of backend requests for laravel.log / Cache key issue

### ISSUE THREAD

**Issue by onlime** [opcodesio/log-viewer#206]
Since having upgraded log-viewer to v2 in a Laravel 10 project, I am experiencing the following issue (also on latest log-viewer v2.2.0):
When clicking on `laravel.log` in the file tree, log-viewer immediately starts to fill up the log with such messages and fires 12-13 requests/s to the backend (3 log lines per request, resulting in more than 36 lines/s):

```
[2023-03-04 17:48:12] local.INFO: Cache key lv:v2.2.0:file:560e00c4-laravel.log:ecf8427e:chunk:3 written with TTL 604800s {"bytes":192199,"tags":[]} 
[2023-03-04 17:48:12] local.INFO: Cache key lv:v2.2.0:file:560e00c4-laravel.log:ecf8427e:metadata written with TTL 604800s {"bytes":1285,"tags":[]} 
[2023-03-04 17:48:12] local.INFO: Cache key lv:v2.2.0:file:560e00c4-laravel.log:metadata written with TTL 604800s {"bytes":368,"tags":[]} 
```

This is not happening when accessing other log files, only on `laravel.log`. I had `CACHE_DRIVER=redis` configured, but same also happens with `CACHE_DRIVER=file`.

```ini
APP_ENV=local
APP_DEBUG=true
CACHE_DRIVER=file
#CACHE_DRIVER=redis
```

If I look at at the XHR requests in browser dev tools network tab, they are all the same GET requests:

`GET /log-viewer/api/logs?file=560e00c4-laravel.log&direction=desc&query=&per_page=25&levels[]=debug&levels[]=info&levels[]=notice&levels[]=warning&levels[]=error&levels[]=critical&levels[]=alert&levels[]=emergency&levels[]=processing&levels[]=processed&levels[]=failed&levels[]=&shorter_stack_traces=false`

log-viewer keeps hammering the backend with such requests even if I deselect all severity levels:

`GET /log-viewer/api/logs?file=560e00c4-laravel.log&direction=desc&query=&per_page=25&levels=none&shorter_stack_traces=false`

I also cleared Local Storage and all cookies, but the problem still persists.

**Comment by arukompas** [comment]
@onlime hey, thanks for the report!

Wow, that's interesting. Log Viewer doesn't have any auto-refresh mechanism so it can't be because of additional log entries in your `laravel.log`.

1. What do these XHR requests return? Are they successful responses with logs, or do they contain errors?
2. Are you searching for something globally (across all logs) when this happens?
3. What's your setup like? Windows/Mac, Valet/Xampp/Docker/Vagrant/etc?
4. What are those cache logs you've posted here? Is that something you've been logging before upgrading Log Viewer, or only added them now to showcase the problem?

Thanks!

**Comment by onlime** [comment]
Thanks @arukompas for following up

> * What do these XHR requests return? Are they successful responses with logs, or do they contain errors?

The are successful HTTP 200 responses and return the following: https://gist.github.com/onlime/69143af1280f3079110d98e6f6519958

> * Are you searching for something globally (across all logs) when this happens?

No, I am not searching for anything and I also did a full reset of all Local Storage data and cookies. I am just clicking on `laravel.log` and the log starts filling up with those 3 log lines per request.

> * What's your setup like? Windows/Mac, Valet/Xampp/Docker/Vagrant/etc?

macOS Ventura, PHP 8.2.3 (Homebrew), running natively (no virtualization), not using any Apache/Nginx webserver, just plaine `php artisan serve` (and `npm run dev` with VITE v4.1.4 for the frontend. And it's a Laravel Jetstream setup with Inertia.js/Vue3, Laravel v10.2.0

> * What are those cache logs you've posted here? Is that something you've been logging before upgrading Log Viewer, or only added them now to showcase the problem?

I've got no idea, they must come from log-viewer itself, I guess. I have also completely removed `laravel.log` but it still pops up as soon as I click on it in log-viewer's file tree.

**Comment by onlime** [comment]
hey, I think I have found the cause of this problem - at nighttime in my brain, will look into the code (missing some lines in my brain…) and explain tomorrow.

**Comment by onlime** [comment]
OK, let me try to explain... Sorry about not noticing earlier that those log lines were triggered by my own implementation. At least in local environment (I am aware that in prod that would be a performance killer), I want to keep track of cache keys written, by writing such messages to the general log: `Log::info("Cache key {$event->key} written with TTL {$event->seconds}s", ['bytes' => mb_strlen(serialize($event->value)), 'tags' => $event->tags])`. This is done by registering `LogKeyWritten` listener in `EventServiceProvider`:

```php
    protected $listen = [
        \Illuminate\Cache\Events\KeyWritten::class => [
            LogKeyWritten::class,
        ],
```

If I turn this off, the problem goes away. But the question is now: What triggers the endless loop of requests in Log Viewer, when those `Cache key lv:v2.2.0:file:* written ...` lines are written to the log?
You are generating those 3 cache keys in `Opcodes\LogViewer\Concerns\LogFile::indexCacheKeyForQuery()` and `LogFile::metadataCacheKey()` in the same request:

```
lv:v2.2.0:file:560e00c4-laravel.log:ecf8427e:chunk:3
lv:v2.2.0:file:560e00c4-laravel.log:ecf8427e:metadata
lv:v2.2.0:file:560e00c4-laravel.log:metadata
```

So how come Log Viewer notices that the underlying log file `laravel.log` was updated (the added 3 log lines) by its own request to the backend and then tries to refresh the whole view? As I understood Log Viewer, it's not actively monitoring the current log file for changes. Just tested this by writing a log line from another session (via Tinker) to laravel.log and it didn't automatically show up in Log Viewer.

Hope this gets us a step further and you can either tell me what I am doing wrong in my cache key logging implementation or in your code that refreshes the view. Thanks!

**Comment by arukompas** [comment]
@onlime wow, what a great explanation and this helps me **a lot**, I can actually reproduce this myself and I get the same result (tons of requests in a neverending loop).

I know why this happens, I'll just need to figure out how to handle this edge case nicely.

When Log Viewer finishes reading and indexing a file (let's say there are 100 logs in the file), it then writes that index to the Cache. **but**, the listener for these cache events then writes another 3 log entries, so now there's 103 logs in the file, but only 100 have been scanned and added to the index. Log Viewer then returns back to the UI with an attribute of `hasMoreResults: true` because the file has more logs than the 100 that has been scanned. 

Because of `hasMoreResults: true` response, the frontend then fires another request to get/index the "remaining" logs, which then does exactly the same - scans the 103 log entries, indexes them, 3 more logs get added on top, and again `hasMoreResults: true` is returned.

---

My first and easiest suggestion would be to add a conditional to the `LogKeyWritten` listener:

```php
if (request()->routeIs('log-viewer.*')) return;
```

This would then exit early without logging the CacheWritten result if the request comes from the Log Viewer. But that might not be ideal if you also want to account for the cache being used by Log Viewer. So I'll see what I can do about it and hopefully it'll be an easy fix :) I'll let you know.

**Comment by arukompas** [comment]
Hey @onlime , I've tagged a new release, `v2.2.1` which includes the fix. It will not refresh itself if the logs have been added the last second (after the scan was complete).

Hope this helps, and thanks again for such thorough explanation of the bug!

**Comment by onlime** [comment]
Wow @arukompas - thank YOU for this fast response and great solution. I wanted to tell you I am perfectly happy with your `if (request()->routeIs('log-viewer.*'))` workaround (and only wanted to raise this issue so that others don't step into this), and then boom, you have already released `v2.2.1` that fixes it!

Thanks so much for this great package you are developing with so much enthusiasm!

**Comment by arukompas** [comment]
This keeps me going. Thanks and I'm glad you enjoy the package ❤️
